Implement tower layer to propagate trace context

hatchan opened the pull request:
Also return the trace id as part of the response

Add trace context to outgoing request going through the API client

Note: The client part is not working as expected. Since it is a shortlived process it won't have time to send any traces that it has buffered. This should be taken care of by `shutdown_tracer_provider` but it doesn't seem to work. Putting a hold on investigating this for now.

arendjr commented on fiberplane-charts/src/Mondrian/generator/utils/getYAxisForConstantValue.ts:
```diff
@@ -26,6 +40,47 @@ export function getYAxisForConstantValue(value: number): Axis {
+    currentTick += interval;
+  }
+
+  return;
```
This `return` seems a bit superfluous :)

arendjr reviewed (approved):
Very nice! :+1:

brettimus commented:
> Btw

yeah! i'll tidy this up then merge into yours

hatchan committed:
Implement tower layer to propagate trace context

Also return the trace id as part of the response

Add trace context to outgoing request going through the API client

hatchan committed:
Move OtelTraceLayer into otel_util

hatchan committed:
Add comments to Client

hatchan committed:
Replace eprintln! with trace!

Remove obsolete instrument macros

hatchan committed:
Resolve clippy
